### PR TITLE
Add heartbeat support for mqtt plugin

### PR DESCRIPTION
**Pull request by gc87**

### CODE REVIEW

File: src/parser/neu_json_mqtt.c [hunk 1]
```diff
@@ -85,3 +86,64 @@ int neu_json_encode_mqtt_resp(void *json_object, void *param)
...
+    int                ret  = 0;
+    neu_json_states_t *resp = (neu_json_states_t *) param;
+
+    void *                 state_array = neu_json_array();
+    neu_json_node_state_t *states      = resp->states;
+    for (int i = 0; i < resp->n_state; i++) {
+        neu_json_elem_t state_elems[] = { {
+                                              .name      = "node",
+                                              .t         = NEU_JSON_STR,
+                                              .v.val_str = (&states[i])->node,
```
Comment (fengzeroz) [inline src/parser/neu_json_mqtt.c:122]: states[i].node

File: src/parser/neu_json_mqtt.c [hunk 2]
```diff
@@ -85,3 +86,64 @@ int neu_json_encode_mqtt_resp(void *json_object, void *param)
...
+                                          },
+                                          {
+                                              .name      = "link",
+                                              .t         = NEU_JSON_INT,
+                                              .v.val_int = (&states[i])->link,
```
Comment (fengzeroz) [inline src/parser/neu_json_mqtt.c:127]: simplify

File: src/parser/neu_json_mqtt.c [hunk 3]
```diff
@@ -85,3 +86,64 @@ int neu_json_encode_mqtt_resp(void *json_object, void *param)
...
+                                          },
+                                          {
+                                              .name = "running",
+                                              .t    = NEU_JSON_INT,
+                                              .v.val_int =
+                                                  (&states[i])->running,
```
Comment (fengzeroz) [inline src/parser/neu_json_mqtt.c:133]: simplify

File: src/connection/mqtt_c_client.c [hunk 4]
```diff
@@ -274,6 +274,7 @@ static BIO *bio_tcp_create(const char *addr, const char *port)
         return NULL;
     }
 
+    BIO_clear_flags(bio, BIO_FLAGS_SHOULD_RETRY);
```
Comment (fengzeroz) [inline src/connection/mqtt_c_client.c:277]: remove

File: plugins/mqtt/util.c [hunk 5]
```diff
@@ -92,24 +97,26 @@ int mqtt_option_init(char *config, neu_mqtt_option_t *option)
...
+    neu_json_elem_t id        = { .name = "client-id", .t = NEU_JSON_STR };
+    neu_json_elem_t upload    = { .name = "upload-topic", .t = NEU_JSON_STR };
+    neu_json_elem_t heartbeat = { .name = "heartbeat-topic",
+                                  .t    = NEU_JSON_STR };
+    neu_json_elem_t format    = { .name = "format", .t = NEU_JSON_INT };
+    neu_json_elem_t ssl       = { .name = "ssl", .t = NEU_JSON_BOOL };
+    neu_json_elem_t host      = { .name = "host", .t = NEU_JSON_STR };
+    neu_json_elem_t port      = { .name = "port", .t = NEU_JSON_INT };
+    neu_json_elem_t username  = { .name = "username", .t = NEU_JSON_STR };
+    neu_json_elem_t password  = { .name = "password", .t = NEU_JSON_STR };
```
Comment (fengzeroz) [inline plugins/mqtt/util.c:111]: optional

File: plugins/mqtt/util.c [hunk 6]
```diff
@@ -92,24 +97,26 @@ int mqtt_option_init(char *config, neu_mqtt_option_t *option)
...
+    neu_json_elem_t ca        = { .name = "ca", .t = NEU_JSON_STR };
+    neu_json_elem_t cert      = { .name = "cert", .t = NEU_JSON_STR };
+    neu_json_elem_t key       = { .name = "key", .t = NEU_JSON_STR };
+    neu_json_elem_t keypass   = { .name = "keypass", .t = NEU_JSON_STR };
```
Comment (fengzeroz) [inline plugins/mqtt/util.c:115]: depend on ssl

File: plugins/mqtt/mqtt.c [hunk 7]
```diff
@@ -510,6 +615,64 @@ static neu_err_code_e trans_data(neu_plugin_t *plugin, void *data)
 //     return NEU_ERR_SUCCESS;
 // }
 
+static neu_err_code_e node_list_get(neu_plugin_t *      plugin,
+                                    neu_reqresp_head_t *head, void *data)
```
Comment (fengzeroz) [inline plugins/mqtt/mqtt.c:619]: node_list_get_state

File: plugins/mqtt/command/heartbeat.c [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    if (0 < json.n_state) {
+        json.states = calloc(json.n_state, sizeof(neu_json_node_state_t));
+    }
+
+    for (int i = 0; i < json.n_state; i++) {
+        struct node_state **nsp = utarray_eltptr(states, (unsigned int) i);
+        struct node_state * ns  = (*nsp);
+        json.states[i].node     = ns->node;
+        json.states[i].link     = ns->link;
+        json.states[i].running  = ns->running;
```
Comment (fengzeroz) [inline plugins/mqtt/command/heartbeat.c:46]: utarray_foreach

File: plugins/mqtt/mqtt.c [hunk 9]
```diff
@@ -198,6 +218,64 @@ static void mqtt_routine_state(void *context, int state)
...
+        cmd.type = NEU_NA_TYPE_APP;
+        int ret  = neu_plugin_op(routine->plugin, header, &cmd);
+        if (ret != 0) { }
+        cmd.type = NEU_NA_TYPE_DRIVER;
+        ret      = neu_plugin_op(routine->plugin, header, &cmd);
+        if (ret != 0) { }
+
+        nng_aio_set_timeout(aio, 3000);
+        nng_aio_defer(aio, mqtt_routine_heartbeat, routine);
+
```
Comment (fengzeroz) [inline plugins/mqtt/mqtt.c:243]: replace with timer event

File: plugins/mqtt/mqtt.c [hunk 10]
```diff
@@ -510,6 +615,64 @@ static neu_err_code_e trans_data(neu_plugin_t *plugin, void *data)
...
+    {
+        // plog_info(plugin, "%s, %s", info->node, info->plugin);
+        neu_reqresp_head_t       header = { 0 };
+        neu_req_get_node_state_t cmd    = { 0 };
+        header.ctx                      = strdup(info->node);
+        header.type                     = NEU_REQ_GET_NODE_STATE;
+
+        strcpy(cmd.node, info->node);
+        int ret = neu_plugin_op(plugin, header, &cmd);
+        if (ret != 0) { }
```
Comment (fengzeroz) [inline plugins/mqtt/mqtt.c:634]: ?

File: plugins/mqtt/mqtt_util.c [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+    if (0 != ret) {
+        free(upload.v.val_str);
+    } else {
+        option->upload_topic = upload.v.val_str;
+    }
+
+    // heartbeat-topic, optional
+    ret = neu_parse_param(config, &error, 1, &heartbeat);
+    if (0 != ret) {
+        free(heartbeat.v.val_str);
```
Comment (fengzeroz) [inline plugins/mqtt/mqtt_util.c:141]: free error

File: plugins/mqtt/command/heartbeat.c [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+
+    int index = 0;
+    utarray_foreach(states, struct node_state **, state_p)
+    {
+        struct node_state *ns      = (*state_p);
+        json.states[index].node    = ns->node;
+        json.states[index].link    = ns->link;
+        json.states[index].running = ns->running;
+        index++;
+    }
```
Comment (fengzeroz) [inline plugins/mqtt/command/heartbeat.c:48]: remove index, use utarray_eltidx

File: plugins/mqtt/command/heartbeat.c [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+{
+    char *                 version  = NEURON_VERSION;
+    neu_json_states_head_t header   = { .version   = version,
+                                      .timpstamp = plugin->common.timestamp };
+    neu_json_states_t      json     = { 0 };
+    char *                 json_str = NULL;
+    json.n_state                    = utarray_len(states);
+    if (0 < json.n_state) {
+        json.states = calloc(json.n_state, sizeof(neu_json_node_state_t));
+    }
```
Comment (fengzeroz) [inline plugins/mqtt/command/heartbeat.c:38]: n_state >= 0

Comment (fengzeroz) [reply plugins/mqtt/command/heartbeat.c:38]: remove if

File: plugins/mqtt/mqtt_util.c [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+    neu_json_elem_t ca        = { .name = "ca", .t = NEU_JSON_STR };
+    neu_json_elem_t cert      = { .name = "cert", .t = NEU_JSON_STR };
+    neu_json_elem_t key       = { .name = "key", .t = NEU_JSON_STR };
+    neu_json_elem_t keypass   = { .name = "keypass", .t = NEU_JSON_STR };
+
+    // client-id, required
+    ret = neu_parse_param(config, &error, 1, &id);
+    if (0 != ret) {
+        plog_error(plugin, "can't decode client-id from json");
+        free(id.v.val_str);
```
Comment (fengzeroz) [inline plugins/mqtt/mqtt_util.c:124]: error

File: plugins/mqtt/mqtt_util.c [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+        return -1;
+    } else {
+        option->clientid = id.v.val_str;
+    }
+
+    // upload-topic, optional
+    ret = neu_parse_param(config, &error, 1, &upload);
+    if (0 != ret) {
+        free(upload.v.val_str);
```
Comment (fengzeroz) [inline plugins/mqtt/mqtt_util.c:133]: error

File: plugins/mqtt/mqtt_util.c [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+    if (0 != ret) {
+        free(heartbeat.v.val_str);
+    } else {
+        option->heartbeat_topic = heartbeat.v.val_str;
+    }
+
+    // format, required
+    ret = neu_parse_param(config, &error, 1, &format);
+    if (0 != ret) {
+        option->format = 0;
```
Comment (fengzeroz) [inline plugins/mqtt/mqtt_util.c:149]: error

Comment (fengzeroz) [reply plugins/mqtt/mqtt_util.c:149]: return error if the required field does not exist

File: plugins/mqtt/mqtt_util.c [hunk 17]
```diff
@@ -0,0 +1,249 @@
...
+            } else {
+                option->ca = ca.v.val_str;
+            }
+
+            // cert, optional
+            ret = neu_parse_param(config, &error, 1, &cert);
+            if (0 == ret) {
+                free(cert.v.val_str);
+            } else {
+                option->cert = cert.v.val_str;
```
Comment (fengzeroz) [inline plugins/mqtt/mqtt_util.c:180]: relatedd fields can be parsed at once

File: plugins/mqtt/mqtt_util.c [hunk 18]
```diff
@@ -0,0 +1,249 @@
...
+            }
+        }
+    }
+
+    // host, required
+    ret = neu_parse_param(config, &error, 1, &host);
+    if (0 != ret) {
+        plog_error(plugin, "can't decode host from json");
+        free(host.v.val_str);
+        return -5;
```
Comment (fengzeroz) [inline plugins/mqtt/mqtt_util.c:208]: should return NEU_ERR_NODE_SETTING_INVALID

File: plugins/mqtt/mqtt_util.c [hunk 19]
```diff
@@ -0,0 +1,249 @@
...
+    } else {
+        option->host = host.v.val_str;
+    }
+
+    // port, required
+    ret = neu_parse_param(config, &error, 1, &port);
+    if (0 != ret) {
+        plog_error(plugin, "can't decode port from json");
+        return -6;
```
Comment (fengzeroz) [inline plugins/mqtt/mqtt_util.c:217]: should return NEU_ERR_NODE_SETTING_INVALID

File: plugins/mqtt/mqtt_util.c (hunk 17 above)
Comment (gc87) [reply plugins/mqtt/mqtt_util.c]: It would be better to keep the status quo here.

File: plugins/mqtt/mqtt_util.c (hunk 18 above)
Comment (gc87) [reply plugins/mqtt/mqtt_util.c:214]: Return NEU_ERR_NODE_SETTING_INVALID in the outer function uniformly

File: plugins/mqtt/mqtt_util.c (hunk 19 above)
Comment (gc87) [reply plugins/mqtt/mqtt_util.c:223]: Return NEU_ERR_NODE_SETTING_INVALID in the outer function uniformly

**Pull request merged**